nfc/NfcCx/Cx/NfcCxRF.cpp

```diff
@@ -3074,6 +3161,146 @@ NfcCxRFInterfaceTargetPresenceCheck(
+    pphNfc_sSeTransceiveInfo_t pSeTransceiveInfo = &RFInterface->pSeTransceiveInfo;
+
+    TRACE_FUNCTION_ENTRY(LEVEL_VERBOSE);
+
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_SE; i++) {
+        if (RFInterface->pLibNfcContext->SEList[i].eSE_Type == phLibNfc_SE_Type_eSE) {
+            hSE_handle = RFInterface->pLibNfcContext->SEList[i].hSecureElement;
+            TRACE_LINE(LEVEL_VERBOSE, "Libnfc based eSE handle is %p", hSE_handle);
+        }
+    }
+
+    LibNfcContext.RFInterface = RFInterface;
+    LibNfcContext.Sequence = RFInterface->pSeqHandler;
+
+    pSeTransceiveInfo->sRecvData.buffer = RFInterface->pSeTransceiveInfo.sRecvData.buffer;
+    pSeTransceiveInfo->sSendData.buffer = RFInterface->pSeTransceiveInfo.sSendData.buffer;
+    pSeTransceiveInfo->sRecvData.length = RFInterface->pSeTransceiveInfo.sRecvData.length;
+    pSeTransceiveInfo->sSendData.length = RFInterface->pSeTransceiveInfo.sSendData.length;
+    RFInterface->pSeTransceiveInfo.timeout = DEFAULT_HCI_TX_RX_TIME_OUT;
+    pSeTransceiveInfo->timeout = DEFAULT_HCI_TX_RX_TIME_OUT;
+
+    nfcStatus = phLibNfc_eSE_Transceive(hSE_handle,
+                                        pSeTransceiveInfo,
+                                        NfcCxRFInterfaceEmbeddedSEAPDUCB,
+                                        &LibNfcContext);
+
+    Status = NfcCxNtStatusFromNfcStatus(nfcStatus);
+    TRACE_FUNCTION_EXIT_NTSTATUS(LEVEL_VERBOSE, Status);
+    return Status;
+}
+
+static VOID
+NfcCxRFInterfaceEmbeddedSEGetATRStringCB(
+    _In_ void* pContext,
+    _In_ pphNfc_sSeAtrInfo_t pResAtrInfo,
+    _In_ NFCSTATUS NfcStatus
+    )
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    PNFCCX_RF_INTERFACE rfInterface = ((PNFCCX_RF_LIBNFC_REQUEST_CONTEXT)pContext)->RFInterface;
+
+    TRACE_FUNCTION_ENTRY(LEVEL_VERBOSE);
+
+    UNREFERENCED_PARAMETER(pResAtrInfo);
+
+    status = NfcCxNtStatusFromNfcStatus(NfcStatus);
+
+    if (NT_SUCCESS(status)) {
+        TRACE_LINE(LEVEL_INFO, "Get ATR was successful. Length = %u", rfInterface->pSeATRInfo.dwLength);
+    }
+    else {
+        TRACE_LINE(LEVEL_ERROR, "Failed to get ATR %!STATUS!", status);
+    }
+
+    NfcCxInternalSequence(rfInterface, ((PNFCCX_RF_LIBNFC_REQUEST_CONTEXT)pContext)->Sequence, status, NULL, NULL);
+
+    TRACE_FUNCTION_EXIT(LEVEL_VERBOSE);
+}
+
+NTSTATUS
+NfcCxRFInterfaceESEGetATRString(
+    _In_ PNFCCX_RF_INTERFACE RFInterface,
+    _In_ NTSTATUS Status,
+    _In_opt_ VOID* /*Param1*/,
+    _In_opt_ VOID* /*Param2*/
+    )
+{
+    NFCSTATUS nfcStatus = NFCSTATUS_SUCCESS;
+    static NFCCX_RF_LIBNFC_REQUEST_CONTEXT LibNfcContext;
+    pphNfc_sSeAtrInfo_t pSeATRInfo = &RFInterface->pSeATRInfo;
+    phLibNfc_Handle hSE_handle = NULL;
+
+    TRACE_FUNCTION_ENTRY(LEVEL_VERBOSE);
+
+    LibNfcContext.RFInterface = RFInterface;
+    LibNfcContext.Sequence = RFInterface->pSeqHandler;
+
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_SE; i++) {
+        if (RFInterface->pLibNfcContext->SEList[i].eSE_Type == phLibNfc_SE_Type_eSE) {
```

Comment:
the constant index cannot be used for eSE because there could be a platform with eSE only. In that case enumerating with constant index will lead to a wrong value